src/js/libcs/OpDrawing.js

```diff
@@ -520,168 +534,131 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
-            var dist0 = Math.abs(center - sol[0]);
-            var dist1 = Math.abs(center - sol[1]);
-            coord = sol[dist0 < dist1 ? 0 : 1];
-            boundary.push(mkp((1 - axis) * coord + x, axis * coord + y));
-        } else { // we need zero or two points of intersection
-            if (sol === null) // have zero intersections
-                return true;
-            coord = 0.5 * (sol[0] + sol[1]);
-            if (!(coord > 0 && coord < extent))
-                return true;
-            var signMid = sign((1 - axis) * coord + x, axis * coord + y);
-            if (signMid === sign1) // intersections outside segment
-                return true;
-            if (isNaN(signMid))
-                return true;
-            // Have two points of solution
-            sort = sort * (sol[1] - sol[0]) > 0 ? 0 : 1;
-            coord = sol[sort];
-            boundary.push(mkp((1 - axis) * coord + x, axis * coord + y));
-            coord = sol[1 - sort];
-            boundary.push(mkp((1 - axis) * coord + x, axis * coord + y));
+
+    // For ovals compute the roots of the x and y discriminant for
+    // the horizontal and vertical tangent points respectively
+    var flats = k00 <= 0 ? [null, null] : [
+        solveRealQuadratic(k00, -2 * k10, k20),
+        solveRealQuadratic(k00, -2 * k01, k02)
+    ];
+
+    var points = [];
+    // Inside state of first corner at top left
+    var bInside = (sign(0, 0) < 0) === (det < 0);
+
+    function doBoundary(sol, other, vert, sort, extent) {
+        // Edge cannot be inside if there are no solutions.
+        var t;
+        if (sol !== null) {
+            var coord, pt;
+            var extent0 = sort > 0 ? 0 : extent;
+            var extent1 = sort > 0 ? extent : 0;
+            // Only include each first corner that falls within the conic
+            if (bInside) {
+                t = "corner";
+                coord = extent0;
+                pt = vert ? mkp(other, coord, t) : mkp(coord, other, t);
+                points.push(pt);
+            }
+            coord = sort > 0 ? sol[0] : sol[1];
+            if (0 < coord && coord < extent || coord === extent1) {
+                bInside = !bInside; // toggles at an intersection
+                t = bInside ? "begin" : "end";
+                pt = vert ? mkp(other, coord, t) : mkp(coord, other, t);
+                points.push(pt);
+            }
+            coord = sort > 0 ? sol[1] : sol[0];
+            if (0 < coord && coord < extent || coord === extent0) {
+                bInside = !bInside; // toggles at an intersection
+                t = bInside ? "begin" : "end";
+                pt = vert ? mkp(other, coord, t) : mkp(coord, other, t);
+                points.push(pt);
+            }
+        } else {
+            var x, y;
+            var flat = flats[vert ? 0 : 1];
+            if (flat !== null) {
+                t = "split";
+                flat = flat[other === 0 ? 0 : 1];
```

Comment:
Using *either* the tangency points *or* the intersection points is an interesting approach which avoids some case distinctions and the sorting issues I had to worry with. :+1: